<a href="https://colab.research.google.com/github/malh718/datasci_2_manipulation/blob/main/507hw2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 507 HW 2

## Loading the dataset


In [ ]:
import pandas as pd
import numpy as np
import re

In [144]:

df=pd.read_csv('https://raw.githubusercontent.com/hantswilliams/HHA_507_2023/main/WK2/data/healthcare_data_cleaning.csv')
df

Patient Age  Gender  City of Residence State of Residence  \
0               45   Other         West Brian           Kentucky   
1               25  Female   East Jocelynfurt           Colorado   
2               51   Other  South Lindseyland               Ohio   
3               18  Female         Taylorfort       Pennsylvania   
4               48  Female          Birdmouth            Montana   
...            ...     ...                ...                ...   
104995          74   Other       West Anthony           Oklahoma   
104996          37    Male          Mckeetown               Utah   
104997          51  Female   Lake Lisachester           Nebraska   
104998     missing  Female       Heidiborough            missing   
104999          45  Female      South Jeffrey      Massachusetts   

       Has Insurance Visited Last Month Payment Method Preferred Doctor  \
0                Yes                Yes      Insurance     Dr. Williams   
1                 No                Yes           Card      Dr. Johnson   
2                Yes                 No           Cash     Dr. Williams   
3                Yes            missing           Cash     Dr. Williams   
4                 No                 No      Insurance        Dr. Brown   
...              ...                ...            ...              ...   
104995           Yes                 No           Card     Dr. Williams   
104996           Yes                 No           Cash        Dr. Jones   
104997           Yes                Yes      Insurance     Dr. Williams   
104998           Yes                 No           Cash        Dr. Smith   
104999            No                Yes           Cash        Dr. Brown   

       Disease Diagnosed Medication Prescribed Type of Appointment  \
0                    Flu                 Med_E             General   
1                    Flu                 Med_B          Specialist   
2               Covid-19                 Med_D             General   
3                    Flu                 Med_A             General   
4                   None                 Med_C          Specialist   
...                  ...                   ...                 ...   
104995          Covid-19                 Med_E          Specialist   
104996          Covid-19                 Med_C             General   
104997           Allergy                 Med_D           Follow-Up   
104998          Covid-19                 Med_A           Follow-Up   
104999           Allergy                 Med_D             General   

       Average Heart Rate Average BP Height (in cm) Weight (in kg)  \
0                      66        111            183             70   
1                      59    missing            174             58   
2                      79        119            161             56   
3                      99        115            171             52   
4                      93         89            186             87   
...                   ...        ...            ...            ...   
104995                 65        112            169             61   
104996            missing        111            165             67   
104997                 89        118            179             71   
104998                 74        123            167             75   
104999                 77        140            167             62   

           Payment Due ($) Last Visit (days ago) Visit Duration (mins)  \
0        65.45721578126224                   193                    92   
1       430.68365678679174                   195               missing   
2        315.0709305262176               missing                    37   
3        320.2998987723972                   175                   108   
4        264.2147372473319                   188                   113   
...                    ...                   ...                   ...   
104995  309.16444606680665                   183                    10   
104996         

In [145]:
data_types = df.dtypes
print(data_types)

Patient Age              object
Gender                   object
City of Residence        object
State of Residence       object
Has Insurance            object
Visited Last Month       object
Payment Method           object
Preferred Doctor         object
Disease Diagnosed        object
Medication Prescribed    object
Type of Appointment      object
Average Heart Rate       object
Average BP               object
Height (in cm)           object
Weight (in kg)           object
Payment Due ($)          object
Last Visit (days ago)    object
Visit Duration (mins)    object
Number of Tests          object
Prescription Cost ($)    object
dtype: object


## Identify and Handle Missing Value

In [171]:

df.replace(["missing"], np.nan, inplace=True)

In [167]:
df.tail(20)

Patient Age  Gender     City of Residence State of Residence  \
104952          74  Female         WestPhilipton           Kentucky   
104957          34  Female         SouthJohntown           Arkansas   
104958          39   Other           Mullinstown            Georgia   
104963          45   Other  NorthJenniferchester            Wyoming   
104964          38    Male         Justinchester         Washington   
104966          59  Female       WestMichaelberg             Oregon   
104972          54    Male             LakeJacob           Missouri   
104974          66  Female             Lopezfort           Delaware   
104977          68    Male         LakePhilipton       NewHampshire   
104978          34    Male             NewNicole            NewYork   
104982          45  Female        LakeAshleyside           Illinois   
104983          59  Female            Brianshire            Georgia   
104985          59   Other            SouthKatie           Michigan   
104987          18  Female             LakeAaron            Vermont   
104988          44    Male           Donaldmouth           Missouri   
104992          44  Female         PortTracistad             Alaska   
104993          42  Female          WestAnnaport            Florida   
104994          53  Female         Michelleburgh        SouthDakota   
104995          74   Other           WestAnthony           Oklahoma   
104997          51  Female       LakeLisachester           Nebraska   

       Has Insurance Visited Last Month Payment Method Preferred Doctor  \
104952            No                Yes      Insurance          DrSmith   
104957            No                Yes           Cash       DrWilliams   
104958           Yes                 No           Cash          DrBrown   
104963            No                Yes           Cash          DrJones   
104964           Yes                 No           Cash        DrJohnson   
104966            No                Yes      Insurance          DrJones   
104972            No                Yes           Cash          DrJones   
104974            No                Yes           Card          DrBrown   
104977           Yes                 No           Cash        DrJohnson   
104978           Yes                 No           Card          DrBrown   
104982           Yes                Yes           Card          DrJones   
104983           Yes                Yes      Insurance        DrJohnson   
104985            No                Yes      Insurance        DrJohnson   
104987           Yes                 No           Card          DrSmith   
104988           Yes                Yes      Insurance          DrJones   
104992           Yes                 No      Insurance          DrBrown   
104993           Yes                 No           Cash        DrJohnson   
104994            No                 No           Card       DrWilliams   
104995           Yes                 No           Card       DrWilliams   
104997           Yes                Yes      Insurance       DrWilliams   

       Disease Diagnosed Medication Prescribed Type of Appointment  \
104952               Flu                  MedE            FollowUp   
104957           Covid19                  MedA          Specialist   
104958               Flu                  MedD            FollowUp   
104963               Flu                  MedD            FollowUp   
104964           Allergy                  MedE          Specialist   
104966              Cold                  MedE          Specialist   
104972           Allergy                  MedD           Emergency   
104974           Allergy                  MedD            FollowUp   
104977               Flu                  MedD          Specialist   
104978           Covid19                  MedA            FollowUp   
104982           Allergy                  MedA          Specialist   
104983               Flu                  MedE            FollowUp   
104985           Allergy     

In [158]:
# Drop rows with any missing values
df.dropna(inplace=True)

In [159]:
# Or if we want to fill missing values for all columns:
df.fillna(0, inplace=True)
df

Patient Age  Gender City of Residence State of Residence Has Insurance  \
0               45   Other         WestBrian           Kentucky           Yes   
7               52    Male      Vanessaburgh        Connecticut           Yes   
8               20  Female   WestBriannabury            Montana            No   
11              40  Female        Jonesmouth        RhodeIsland            No   
12              34    Male       WestDeborah             Oregon           Yes   
...            ...     ...               ...                ...           ...   
104992          44  Female     PortTracistad             Alaska           Yes   
104993          42  Female      WestAnnaport            Florida           Yes   
104994          53  Female     Michelleburgh        SouthDakota            No   
104995          74   Other       WestAnthony           Oklahoma           Yes   
104997          51  Female   LakeLisachester           Nebraska           Yes   

       Visited Last Month Payment Method Preferred Doctor Disease Diagnosed  \
0                     Yes      Insurance       DrWilliams               Flu   
7                     Yes      Insurance       DrWilliams               Flu   
8                     Yes           Card          DrJones               Flu   
11                     No           Card          DrJones           Covid19   
12                    Yes           Card          DrJones           Covid19   
...                   ...            ...              ...               ...   
104992                 No      Insurance          DrBrown              Cold   
104993                 No           Cash        DrJohnson           Covid19   
104994                 No           Card       DrWilliams               Flu   
104995                 No           Card       DrWilliams           Covid19   
104997                Yes      Insurance       DrWilliams           Allergy   

       Medication Prescribed Type of Appointment Average Heart Rate  \
0                       MedE             General                 66   
7                       MedE             General                 99   
8                       MedC          Specialist                 67   
11                      MedE          Specialist                 65   
12                      MedB          Specialist                 76   
...                      ...                 ...                ...   
104992                  MedA            FollowUp                 76   
104993                  MedD          Specialist                 80   
104994                  MedB           Emergency                 68   
104995                  MedE          Specialist                 65   
104997                  MedD            FollowUp                 89   

       Average BP Height (in cm) Weight (in kg)    Payment Due ($)  \
0             111            183             70   6545721578126224   
7             129            170             47  38708504353508863   
8             108            173             86   1725483427968999   
11            106            166             82   2991426761400895   
12            111            176             40   1381128079141103   
...           ...            ...            ...                ...   
104992         93            178             66  46256567087397013   
104993        129            171             78  13135632576282202   
104994        109            172             75   4880316041186948   
104995        112            169             61  30916444606680665   
104997        118            179             71   3208304180033907   

       Last Visit (days ago) Visit Duration (mins) Number of Tests  \
0                        193                    92               5   
7                        179                   108               2   
8                        174                    91               2   
11                       193                    92               1   
12                       182             

In [160]:
df.describe()

Patient Age Gender City of Residence State of Residence Has Insurance  \
count        28666  28666             28666              28666         28666   
unique          83      3             17371                 50             2   
top             18  Other       WestMichael          Minnesota            No   
freq          1243   9686                33                629         14378   

       Visited Last Month Payment Method Preferred Doctor Disease Diagnosed  \
count               28666          28666            28666             28666   
unique                  2              3                5                 4   
top                   Yes           Cash       DrWilliams              Cold   
freq                14348           9647             5803              7228   

       Medication Prescribed Type of Appointment Average Heart Rate  \
count                  28666               28666              28666   
unique                     5                   4                 69   
top                     MedB          Specialist                 80   
freq                    5784                7268               1142   

       Average BP Height (in cm) Weight (in kg)   Payment Due ($)  \
count       28666          28666          28666             28666   
unique         61             71             86             28666   
top            80            176             67  6545721578126224   
freq          792           1153            786                 1   

       Last Visit (days ago) Visit Duration (mins) Number of Tests  \
count                  28666                 28666           28666   
unique                    76                   111              11   
top                      184                    21               4   
freq                    1250                   302            7219   

       Prescription Cost ($)  
count                  28666  
unique                 28666  
top         1571751735532889  
freq                       1

What this code does is it checks to see if there are any columns that are missing values. Since there are "missing" for a number of the values that will be converted into NaN and subsequently dropped. The data went from 105,000 rows to 28,666 rows. With the columns remaning the same at 20.

## Clean Column Names

In [153]:
def clean_value(value):
    cleaned_value = re.sub(r'\s+|[^a-zA-Z0-9]', '', value)
    return cleaned_value

In [154]:
## clean colomn names
for column in df.columns:
    df[column] = df[column].apply(clean_value)

print(df)

       Patient Age  Gender City of Residence State of Residence Has Insurance  \
0               45   Other         WestBrian           Kentucky           Yes   
7               52    Male      Vanessaburgh        Connecticut           Yes   
8               20  Female   WestBriannabury            Montana            No   
11              40  Female        Jonesmouth        RhodeIsland            No   
12              34    Male       WestDeborah             Oregon           Yes   
...            ...     ...               ...                ...           ...   
104992          44  Female     PortTracistad             Alaska           Yes   
104993          42  Female      WestAnnaport            Florida           Yes   
104994          53  Female     Michelleburgh        SouthDakota            No   
104995          74   Other       WestAnthony           Oklahoma           Yes   
104997          51  Female   LakeLisachester           Nebraska           Yes   

       Visited Last Month P

I made sure to add import re to my first code cell, to ensure that the built in package would be loaded in and I would be able to "clean the columns." What this code above does is it takes away any white spaces from the column names as well as deletes any special characters.

In [155]:
df.sample(10)


Patient Age  Gender  City of Residence State of Residence  \
60084           43  Female      EastStephanie       Pennsylvania   
61378           44    Male  NorthKennethmouth         California   
12              34    Male        WestDeborah             Oregon   
46326           33   Other         LakeAmanda            Georgia   
100911          34    Male         PortAustin             Kansas   
23602           52  Female         Howellbury               Utah   
54114           47    Male    EastTheresafort       NewHampshire   
16501           63    Male      PortEdwardton            Montana   
51834           18  Female         Shawnshire          NewMexico   
23393           18    Male         Debraville           Missouri   

       Has Insurance Visited Last Month Payment Method Preferred Doctor  \
60084             No                Yes           Cash        DrJohnson   
61378            Yes                Yes           Card        DrJohnson   
12               Yes                Yes           Card          DrJones   
46326             No                 No      Insurance          DrSmith   
100911           Yes                 No      Insurance          DrBrown   
23602             No                Yes           Card          DrJones   
54114            Yes                Yes           Cash       DrWilliams   
16501             No                Yes      Insurance          DrBrown   
51834             No                 No           Cash          DrBrown   
23393             No                 No           Cash          DrSmith   

       Disease Diagnosed Medication Prescribed Type of Appointment  \
60084            Covid19                  MedD           Emergency   
61378            Covid19                  MedC          Specialist   
12               Covid19                  MedB          Specialist   
46326                Flu                  MedD            FollowUp   
100911           Covid19                  MedC           Emergency   
23602            Allergy                  MedA            FollowUp   
54114            Covid19                  MedD           Emergency   
16501            Allergy                  MedA            FollowUp   
51834                Flu                  MedA          Specialist   
23393               Cold                  MedD            FollowUp   

       Average Heart Rate Average BP Height (in cm) Weight (in kg)  \
60084                  69        102            174             66   
61378                  95         97            166             46   
12                     76        111            176             40   
46326                  66        110            181             71   
100911                 77        103            159             91   
23602                  86         94            174             85   
54114                  57        139            174             46   
16501                  84        111            164             90   
51834                  70        104            170             67   
23393                  85         98            186             62   

          Payment Due ($) Last Visit (days ago) Visit Duration (mins)  \
60084   20508322195467642                   169                    90   
61378     471039886201125                   169                    12   
12       1381128079141103                   182                    22   
46326   45782181406313117                   179                    73   
100911   2808403779237914                   167                    14   
23602    2906953032089101                   175                    48   
54114    9672503383538967                   205                   105   
16501    6226136721723952                   156                   101   
51834    3580038487926347                   173                    92   
23393   20875304837931347                   183                    20   

       Number of Tests Prescription Cost ($)  
60084                3      49643820346

## Remove Duplicate Rows and Columns

In [198]:
# Identify and remove duplicate rows
df.drop_duplicates(inplace=True)
df


Patient Age  Gender City of Residence State of Residence Has Insurance  \
0               45   Other         WestBrian           Kentucky           Yes   
7               52    Male      Vanessaburgh        Connecticut           Yes   
8               20  Female   WestBriannabury            Montana            No   
11              40  Female        Jonesmouth        RhodeIsland            No   
12              34    Male       WestDeborah             Oregon           Yes   
...            ...     ...               ...                ...           ...   
104992          44  Female     PortTracistad             Alaska           Yes   
104993          42  Female      WestAnnaport            Florida           Yes   
104994          53  Female     Michelleburgh        SouthDakota            No   
104995          74   Other       WestAnthony           Oklahoma           Yes   
104997          51  Female   LakeLisachester           Nebraska           Yes   

       Visited Last Month Payment Method Preferred Doctor Disease Diagnosed  \
0                     Yes      Insurance       DrWilliams               Flu   
7                     Yes      Insurance       DrWilliams               Flu   
8                     Yes           Card          DrJones               Flu   
11                     No           Card          DrJones           Covid19   
12                    Yes           Card          DrJones           Covid19   
...                   ...            ...              ...               ...   
104992                 No      Insurance          DrBrown              Cold   
104993                 No           Cash        DrJohnson           Covid19   
104994                 No           Card       DrWilliams               Flu   
104995                 No           Card       DrWilliams           Covid19   
104997                Yes      Insurance       DrWilliams           Allergy   

       Medication Prescribed Type of Appointment Average Heart Rate  \
0                       MedE             General                 66   
7                       MedE             General                 99   
8                       MedC          Specialist                 67   
11                      MedE          Specialist                 65   
12                      MedB          Specialist                 76   
...                      ...                 ...                ...   
104992                  MedA            FollowUp                 76   
104993                  MedD          Specialist                 80   
104994                  MedB           Emergency                 68   
104995                  MedE          Specialist                 65   
104997                  MedD            FollowUp                 89   

       Average BP Height (in cm) Weight (in kg)    Payment Due ($)  \
0             111            183             70   6545721578126224   
7             129            170             47  38708504353508863   
8             108            173             86   1725483427968999   
11            106            166             82   2991426761400895   
12            111            176             40   1381128079141103   
...           ...            ...            ...                ...   
104992         93            178             66  46256567087397013   
104993        129            171             78  13135632576282202   
104994        109            172             75   4880316041186948   
104995        112            169             61  30916444606680665   
104997        118            179             71   3208304180033907   

       Last Visit (days ago) Visit Duration (mins) Number of Tests  \
0                        193                    92               5   
7                        179                   108               2   
8                        174                    91               2   
11                       193                    92               1   
12                       182             

This code removes rows that have the same infromation so as to not skew any data. In this case, rows with no information were already deleted, and subsequently there are no duplicated rows.

In [197]:
# Identify and remove duplicate columns
df2=df.T.drop_duplicates().T
df2

Patient Age  Gender City of Residence State of Residence Has Insurance  \
0               45   Other         WestBrian           Kentucky           Yes   
7               52    Male      Vanessaburgh        Connecticut           Yes   
8               20  Female   WestBriannabury            Montana            No   
11              40  Female        Jonesmouth        RhodeIsland            No   
12              34    Male       WestDeborah             Oregon           Yes   
...            ...     ...               ...                ...           ...   
104992          44  Female     PortTracistad             Alaska           Yes   
104993          42  Female      WestAnnaport            Florida           Yes   
104994          53  Female     Michelleburgh        SouthDakota            No   
104995          74   Other       WestAnthony           Oklahoma           Yes   
104997          51  Female   LakeLisachester           Nebraska           Yes   

       Visited Last Month Payment Method Preferred Doctor Disease Diagnosed  \
0                     Yes      Insurance       DrWilliams               Flu   
7                     Yes      Insurance       DrWilliams               Flu   
8                     Yes           Card          DrJones               Flu   
11                     No           Card          DrJones           Covid19   
12                    Yes           Card          DrJones           Covid19   
...                   ...            ...              ...               ...   
104992                 No      Insurance          DrBrown              Cold   
104993                 No           Cash        DrJohnson           Covid19   
104994                 No           Card       DrWilliams               Flu   
104995                 No           Card       DrWilliams           Covid19   
104997                Yes      Insurance       DrWilliams           Allergy   

       Medication Prescribed Type of Appointment Average Heart Rate  \
0                       MedE             General                 66   
7                       MedE             General                 99   
8                       MedC          Specialist                 67   
11                      MedE          Specialist                 65   
12                      MedB          Specialist                 76   
...                      ...                 ...                ...   
104992                  MedA            FollowUp                 76   
104993                  MedD          Specialist                 80   
104994                  MedB           Emergency                 68   
104995                  MedE          Specialist                 65   
104997                  MedD            FollowUp                 89   

       Average BP Height (in cm) Weight (in kg)    Payment Due ($)  \
0             111            183             70   6545721578126224   
7             129            170             47  38708504353508863   
8             108            173             86   1725483427968999   
11            106            166             82   2991426761400895   
12            111            176             40   1381128079141103   
...           ...            ...            ...                ...   
104992         93            178             66  46256567087397013   
104993        129            171             78  13135632576282202   
104994        109            172             75   4880316041186948   
104995        112            169             61  30916444606680665   
104997        118            179             71   3208304180033907   

       Last Visit (days ago) Visit Duration (mins) Number of Tests  \
0                        193                    92               5   
7                        179                   108               2   
8                        174                    91               2   
11                       193                    92               1   
12                       182             

This code removes any duplicated columns, and as seen there are none.

## Create New Column

In [114]:
# Define clinically defined thresholds for categorization
age_high = '25'

# Create a new column "age_Status" based on thresholds
df['age_Status'] = df['Patient Age'].apply(lambda x: 'Later Adult' if x >= age_high else 'young adult')

print(df[['Patient Age', 'Gender', 'age_Status']])


       Patient Age  Gender   age_Status
0               45   Other  Later Adult
7               52    Male  Later Adult
8               20  Female  young adult
11              40  Female  Later Adult
12              34    Male  Later Adult
...            ...     ...          ...
104992          44  Female  Later Adult
104993          42  Female  Later Adult
104994          53  Female  Later Adult
104995          74   Other  Later Adult
104997          51  Female  Later Adult

[28666 rows x 3 columns]


In [115]:

print(df[['Patient Age', 'Gender', 'age_Status']].head(10))


   Patient Age  Gender   age_Status
0           45   Other  Later Adult
7           52    Male  Later Adult
8           20  Female  young adult
11          40  Female  Later Adult
12          34    Male  Later Adult
13          43  Female  Later Adult
15          54  Female  Later Adult
17          64   Other  Later Adult
21          26   Other  Later Adult
22          35   Other  Later Adult


Here, any who is 25 years old or older is considered a Later Adult, and anyone who is younger than 25 is considered a young adult. This column name is referred to as age_Status, and basically just shows if you are a young or later adult.

## Grouby and Pivot Tables

In [193]:
res = df.groupby('State of Residence').agg({
    'Patient Age': 'max'
})
print(res)




                   Patient Age
State of Residence            
Alabama                     86
Alaska                      82
Arizona                     94
Arkansas                    97
California                  90
Colorado                    90
Connecticut                 87
Delaware                    87
Florida                     88
Georgia                     89
Hawaii                      85
Idaho                       82
Illinois                    93
Indiana                     95
Iowa                        99
Kansas                      90
Kentucky                    98
Louisiana                   96
Maine                       86
Maryland                    86
Massachusetts               99
Michigan                    88
Minnesota                   88
Mississippi                 94
Missouri                    94
Montana                     91
Nebraska                    93
Nevada                      95
NewHampshire                97
NewJersey                   96
NewMexic

In [117]:
df['Patient Age'].value_counts()

18    1243
42     802
44     781
47     767
43     747
      ... 
94       3
98       2
97       2
99       2
91       2
Name: Patient Age, Length: 83, dtype: int64

In [174]:
grouped_data = df.groupby('Patient Age').agg({
    'Average BP': 'mean'
})
print(grouped_data)

               Average BP
Patient Age              
100          2.802675e+09
18                    inf
19                    inf
20                    inf
21                    inf
...                   ...
95           1.211168e+25
96           3.002753e+10
97           5.456850e+04
98           6.196500e+03
99           5.606200e+04

[83 rows x 1 columns]


In [187]:
state = df.groupby('Gender').agg({
    'Average BP': 'median'
})
print(state)

        Average BP
Gender            
Female       110.0
Male         109.0
Other        110.0


In [119]:
diagnosis_pivot = df.pivot_table(values='Patient Age', index='Disease Diagnosed', columns='Gender', aggfunc='count')
print(diagnosis_pivot)


Gender             Female  Male  Other
Disease Diagnosed                     
Allergy              2350  2397   2395
Cold                 2396  2417   2415
Covid19              2367  2289   2455
Flu                  2405  2359   2421


## Polars


In [229]:
pip install polars

In [257]:
import polars as pl


In [283]:
import time
start_time = time.time()
df3 = pl.read_csv("healthcare_data_cleaning.csv")

pandas_time = time.time() - start_time
print(pandas_time)
print(f"Pandas Execution Time: {pandas_time:f} seconds")

0.11727523803710938
Pandas Execution Time: 0.117275 seconds
